reconcile/templating/renderer.py

```diff
@@ -0,0 +1,179 @@
+import logging
+import os
+from abc import ABC, abstractmethod
+from collections.abc import Callable
+from typing import Optional
+
+from pydantic import BaseModel
+from ruamel import yaml
+
+from reconcile.gql_definitions.templating.template_collection import (
+    TemplateCollectionV1,
+    TemplateCollectionVariablesV1,
+    TemplateV1,
+    query,
+)
+from reconcile.templating.rendering import (
+    TemplateData,
+    create_renderer,
+)
+from reconcile.utils import gql
+from reconcile.utils.gql import init_from_config
+from reconcile.utils.runtime.integration import (
+    PydanticRunParams,
+    QontractReconcileIntegration,
+)
+
+QONTRACT_INTEGRATION = "template-renderer"
+
+APP_INTERFACE_PATH_SEPERATOR = "/"
+
+
+def get_template_collections(
+    query_func: Optional[Callable] = None,
+) -> list[TemplateCollectionV1]:
+    if not query_func:
+        query_func = gql.get_api().query
+    return query(query_func).template_collection_v1 or []
+
+
+class TemplateOutput(BaseModel):
+    path: str
+    content: str
+
+
+class FilePersistence(ABC):
+    @abstractmethod
+    def write(self, outputs: list[TemplateOutput]) -> None:
+        pass
+
+    @abstractmethod
+    def read(self, path: str) -> Optional[str]:
+        pass
+
+
+class LocalFilePersistence(FilePersistence):
+    def __init__(self, app_interface_data_path: str) -> None:
+        self.app_interface_data_path = app_interface_data_path
+
+    def write(self, outputs: list[TemplateOutput]) -> None:
+        for output in outputs:
+            with open(
+                f"{join_path(self.app_interface_data_path, output.path)}",
+                "w",
+                encoding="utf-8",
+            ) as f:
+                f.write(output.content)
+
+    def read(self, path: str) -> Optional[str]:
+        try:
+            with open(
+                f"{join_path(self.app_interface_data_path, path)}",
+                "r",
+                encoding="utf-8",
+            ) as f:
+                return f.read()
+        except FileNotFoundError:
+            logging.debug(f"File not found: {path}, need to create it")
+
+
+def unpack_static_variables(
+    collection_variables: TemplateCollectionVariablesV1,
+) -> dict:
+    return collection_variables.static or {}
+
+
+def unpack_dynamic_variables(
+    collection_variables: TemplateCollectionVariablesV1, gql: gql.GqlApi
+) -> dict:
+    if not collection_variables.dynamic:
+        return {}
+    return {
+        dv.name: list(gql.query(dv.query).values())
+        for dv in collection_variables.dynamic or []
+    }
+
+
+class TemplateRendererIntegrationParams(PydanticRunParams):
+    app_interface_data_path: Optional[str]
+
+
+def join_path(base: str, sub: str) -> str:
+    # not using os.path.sep, since app-interface relies on unix paths
+    if sub.startswith(APP_INTERFACE_PATH_SEPERATOR):
+        return os.path.join(base, sub[1:])
+    return os.path.join(base, sub)
```

Comment:
Because of your comment above, `os.path.join` is using `os.path.sep` too :)

Reply:
Thought about that. But, what if someone runs it on Windows? Our paths are POSIX, I even thought about raising an exception if this was run on windows.

Reply:
I think no one can run qontract-reconcile on Windows, especially as we rely on other binaries like terraform, git, ... Even on Windows, you can run the qontract-reconcile container.